File: backend/app/schemas/ai.py
"""AI Schemas."""

from datetime import datetime
from typing import Optional, Any
from uuid import UUID

from pydantic import BaseModel, Field


class NaturalLanguageQuery(BaseModel):
    """Schema for natural language query."""
    query: str = Field(..., min_length=1)
    context: Optional[dict[str, Any]] = None


class ParsedIntent(BaseModel):
    """Schema for parsed intent."""
    type: str
    entity: Optional[str] = None
    filters: Optional[dict[str, Any]] = None


class SuggestedAction(BaseModel):
    """Schema for suggested action."""
    action: str
    parameters: dict[str, Any]
    explanation: str


class NaturalLanguageResponse(BaseModel):
    """Schema for natural language response."""
    query: str
    parsed_intent: dict[str, Any]
    explanation: str
    suggested_action: Optional[SuggestedAction] = None
    result_count: Optional[int] = None
    query_id: Optional[str] = None


class QueryFeedback(BaseModel):
    """Schema for query feedback."""
    was_helpful: bool
    feedback: Optional[str] = None


class NarrativeRequest(BaseModel):
    """Schema for narrative request."""
    template: str
    data: dict[str, Any]
    max_length: Optional[int] = 500
    tone: Optional[str] = "professional"


class NarrativeResponse(BaseModel):
    """Schema for narrative response."""
    text: str
    generated_at: Optional[datetime] = None


class SemanticSearchRequest(BaseModel):
    """Schema for semantic search request."""
    query: str
    resource_type: Optional[str] = None
    limit: Optional[int] = 10


class SemanticSearchResponse(BaseModel):
    """Schema for semantic search response."""
    query: str
    results: list[dict[str, Any]]
    total: int

File: frontend/src/types/ai.ts
// AI-related types

export interface NLQueryRequest {
  query: string;
  context?: Record<string, any>;
}

export interface ParsedIntent {
  intent: string;
  entity_type?: string;
  filters: Record<string, any>;
  clarification_needed?: boolean;
  clarification_question?: string;
  type?: string;
}

export interface SuggestedAction {
  action: string;
  parameters: Record<string, any>;
  explanation: string;
}

export interface NLQueryResponse {
  query: string;
  parsed_intent: ParsedIntent;
  explanation: string;
  result_count?: number;
  suggested_api_call?: {
    endpoint: string;
    method: string;
    params: Record<string, any>;
  };
  suggested_action?: SuggestedAction;
}

export interface NarrativeRequest {
  template: string;
  data: Record<string, any>;
  max_length?: number;
  tone?: 'professional' | 'casual' | 'technical';
}

export interface NarrativeResponse {
  text: string;
  generated_at?: string;
  tokens_used?: number;
}

export interface AnomalyAlert {
  id: string;
  resource_type: string;
  resource_id: string;
  record_type?: string;
  record_id?: string;
  anomaly_type: string;
  score: number;
  description: string;
  explanation: string;
  reasons?: string[];
  detected_at: string;
  resolved?: boolean;
}

export interface DocumentExtractionResult {
  document_id: string;
  extracted_text: string;
  entities: ExtractedEntity[];
  structured_data: Record<string, ExtractedField>;
  confidence: number;
}

export interface ExtractedEntity {
  type: string;
  value: string;
  confidence: number;
  start_pos?: number;
  end_pos?: number;
}

export interface ExtractedField {
  value: any;
  confidence: number;
  source?: string;
}

export interface SemanticSearchResult {
  id: string;
  resource_type: string;
  title: string;
  snippet: string;
  score: number;
  metadata?: Record<string, any>;
}

export interface SemanticSearchRequest {
  query: string;
  resource_types?: string[];
  limit?: number;
  offset?: number;
}

export interface ExperienceRecommendation {
  assumption_type: string;
  segment?: string;
  current_value: number;
  recommended_value: number;
  change_percent: number;
  confidence: number;
  sample_size: number;
  rationale: string;
  impact?: {
    reserve_change: number;
    direction: 'increase' | 'decrease';
  };
}

export interface AIFeatureStatus {
  enabled: boolean;
  healthy: boolean;
  features: {
    smart_import: boolean;
    natural_language: boolean;
    anomaly_detection: boolean;
    narrative_generation: boolean;
    semantic_search: boolean;
    document_extraction: boolean;
    experience_recommendations: boolean;
  };
}

export interface AIFeatures {
  smart_import: boolean;
  natural_language: boolean;
  anomaly_detection: boolean;
  narrative_generation: boolean;
  semantic_search: boolean;
  document_extraction: boolean;
  experience_recommendations: boolean;
}

export interface NaturalLanguageQuery {
  query: string;
  context?: Record<string, any>;
}

export interface QueryFeedback {
  query_id: string;
  helpful: boolean;
  feedback?: string;
}

export interface QueryHistoryItem {
  id: string;
  query: string;
  result: NLQueryResponse;
  timestamp: string;
}

export interface ColumnMappingSuggestion {
  source_column: string;
  suggested_field: string;
  confidence: number;
  type?: string;
}

export interface DataQualityIssue {
  row_number: number;
  column: string;
  issue_type: string;
  description: string;
  severity: 'error' | 'warning' | 'info';
}
